learning-another-language: content (#7)
Ada-Developers-Academy/core-unit-3: learning-another-language/hello-world.md

# Hello, World!

## Introduction

A well-known exercise to accomplish when learning another programming language is "Hello, World!" We'll know we're successful if we see the text `Hello, World!` appear somewhere after running the code.

## Running Code in the REPL

To run JavaScript code, we need to consider its _runtime environment_. A **runtime environment** is a set of technologies that determine _how_ to execute the code.

Some examples of runtime environments that can run JavaScript:

- The runtime environment of a web browser like Mozilla Firefox and Google Chrome
- [Node.js](https://nodejs.org/en/)
- REPLs like [replit.com](https://replit.com/)

For this portion of the curriculum, we will use [replit.com](https://replit.com/).

The following steps will let us create and run JavaScript programs with ease:

1. Create an account on [replit.com](https://replit.com/) and log-in
1. Visit [the Hello, World! REPL](https://replit.com/@adacore/hello-world#index.js)

> 

Reviewer: add an image and explanation for the preview we see for a repl that we aren't invited to edit
```suggestion

When we open a REPL that we don't own, we will see a page with the project name, and a play button to run the project. We can toggle the view to show the code. We can also make a copy of the project in our own account by clicking Fork.

![JavaScript readonly REPL. The browser displays the name of the project, hello-world, and a large green button to run the project. A toggle labelled Code and button labelled Fork are also visible.](../assets/learning-a-new-language_hello-world_repl-unowned.png)

Let's Fork this project and see what that lets us do!

```

> On the right side, we see the _console_. We can use this console to execute small, quick JavaScript snippets, such as `2 + 3`, `4 * 5`, or any other valid code.

Reviewer: small connector to transition from the suggested addition above
```suggestion
Now, on the right side, we see the _console_. We can use this console to execute small, quick JavaScript snippets, such as `2 + 3`, `4 * 5`, or any other valid code.
```

| `console`                                           | A variable. The word "console" usually refers to the Terminal, command line, or wherever output logs get printed. |
| `console.log( ... )`                                | A function call that takes in some arguments.                                                                    |
| `'Hello, World!'`                                   | A string literal. This string is surrounded by single-quotes (`'`).                                               |
| `;`                                                 | A semicolon.                                                                                                      |

Let's debrief in slightly more detail.

### The Console

`console` seems a little random here. It's lowercase, like Python variables, but there seems to be no variable assignment. We don't see any `import`s or anything like that, so where did `console` come from?

`console` is a variable that references a specific object: the console! This console object gives us access to wherever the program prints out output.

> It seems that `console` is within scope of our JavaScript file by default.

Reviewer: Be a little more transparent in the thinking that leads to the conclusion of this being something in scope. The existing line sounds like the author isn't quite sure to me.
```suggestion
Since we were able to use it without any setup, we might guess that `console` is in scope in our JavaScript file by default.
```

At this moment, at the beginning of our JavaScript journey, we won't go deeper into this.

### Calling a Function

> `console.log()` is calling a function. We might not know all the details about this function, but we know that it's responsible for printing objects to the console.

Reviewer: I don't know that we _know_ this, but we're pretty sure. Use the exact name to show our thinking on this.
```suggestion
`console.log()` is calling a function. We might not know all the details about this function, but we're pretty sure that it's responsible for logging (printing) objects to the console.
```

We may make guesses that:

1. The function being called here is named `log`

> 1. The function `log` is a member of `console` (hence the dot-operator)

Reviewer: a little more explicit about what to notice
```suggestion
1. The function `log` is a member of `console` (hence the dot-operator to access it)
```

1. We use the dot-operator to access members of objects

> 1. The parens `()` is how we call and invoke the function

Reviewer: numerosity
```suggestion
1. The parens `()` are how we call and invoke the function
```

### Single Quotes for Strings

The string literal `'Hello, World!'` is surrounded by single-quotes, instead of double-quotes.

> If we experiment and replace it with double-quotes (`console.log("Hello, World!");`), we'll see that our code still runs as expected!

Reviewer: expand this out into a full line rather than inline
no suggestion because code
markdown in following comment

Reviewer: If we experiment and replace it with double-quotes

<!-- prettier-ignore-start -->
```js
console.log("Hello, World!");
```
<!-- prettier-ignore-end -->

we'll see that our code still runs as expected!

Author: hahaha thank you for the prettier ignore comments :D

Reviewer: Not really necessary, right? Since prettier wants to make double quotes anyway. Oh well. 🤷

In JavaScript, we can create string literals with either syntax.

### !callout-info

## This Curriculum Will Prefer Single-Quotes

The text in this curriculum will prefer single-quotes. In the grand scheme of things, the most important thing is consistency within a team.

### !end-callout

### Ends With a Semicolon

This line of code ends with a semicolon (`;`).

In JavaScript, semicolons are used to conclude statements.

### !callout-info

## Statements

> In programming, a statement is a unit of code that contains an action to execute. We use statements mainly to _perform an action with side effects_. The following are examples of statements:

Reviewer: Elsewhere in the curriculum, we stress avoiding side effects, so here's a rephrase to avoid possible confusion around the perception of mixed messaging.
```suggestion
In programming, a statement is a unit of code that contains an action to execute. We use statements mainly in the forms of expressions that update values, or control statements that affect the flow of the program. The following are examples of statements:
```

- `foo = 'A value for foo';`
- `return foo;`
- `if` statements

### !end-callout

> Many modern JavaScript flavors, standards, and teams are beginning to prefer _excluding_ semicolons, relying on "Automatic Semicolon Insertion."

Reviewer: gotta represent semicolon gang!
```suggestion
Many modern JavaScript flavors, standards, and teams are beginning to prefer _excluding_ semicolons, relying on "Automatic Semicolon Insertion." But there are still many teams who prefer to _include_ semicolons due to the sometimes surprising effects of Automatic Semicolon Insertion.
```
